### PR TITLE
#727 wip, added tests, without migrations

### PR DESCRIPTION
**Pull request by benzkji**
I've added tests, but don't know how I could add migrations for the new models needed for testing.

### CODE REVIEW

File: modeltranslation/tests/translation.py [hunk 1]
```diff
@@ -85,6 +85,22 @@ class ForeignKeyModelTranslationOptions(TranslationOptions):
...
+    fields = (
+        "test",
+    )
+
+
+@register(models.SecondSetNullForeignKeyModel)
+class SecondSetNullForeignKeyModel(TranslationOptions):
+    fields = (
+        "test",
+    )
```
Comment (last-partizan) [inline modeltranslation/tests/translation.py:99]: I see the problem now.

You should register for translation your base model, if you're using fields from base model.

Please try it. And if it fixes the issue, we can raise error "ImproperlyConfigured()" or something like that.

Comment (benzkji) [reply modeltranslation/tests/translation.py:99]: Yep, that would make it work. If my base model is not under my control, it will not. There probably would be a solution, but the Django code is very complex :|. Let me dig a bit more into...

Comment (benzkji) [reply modeltranslation/tests/translation.py:99]: Sorry. Misunderstood. It's this section: https://django-modeltranslation.readthedocs.io/en/latest/registration.html#translationoptions-fields-inheritance

So I would register the abstract base model with it's foreign key for translation. And maye the sub models, with additional translated fields, right?

Comment (benzkji) [reply modeltranslation/tests/translation.py:99]: Just found I got a third model, inheriting from base class, without having the specific field translated. That would not be possible anymore...but would probably not be a problem, besides having unused fields.

Comment (last-partizan) [reply modeltranslation/tests/translation.py:99]: Main point: you register fields for the model they're defined on.

I'm not sure, if you also need to register child models (with empty fields), try both variants and see which works.

Comment (benzkji) [reply modeltranslation/tests/translation.py:99]: I've done this many times without problems (registering only the final model, but not the abstract base class). But yeah, I can see why it doesn't work in my edge case, with foreign key.

**Comment by benzkji** [comment]
I can't get it to work properly. See my latest commit, that's how it's supposed to be setup?

File: modeltranslation/tests/translation.py [hunk 2]
```diff
@@ -85,6 +85,26 @@ class ForeignKeyModelTranslationOptions(TranslationOptions):
...
 
 
+@register(models.SetNullForeignKeyBaseModel)
+class SetNullForeignKeyBaseModelTranslationOptions(TranslationOptions):
+    fields = [
+        "test",
+    ]
+
+@register(models.FirstSetNullForeignKeyModel)
+class FirstSetNullForeignKeyModelTranslationOptions(SetNullForeignKeyBaseModelTranslationOptions, TranslationOptions):
```
Comment (last-partizan) [inline modeltranslation/tests/translation.py:95]: No need to inherit previous class:

image

Comment (benzkji) [reply modeltranslation/tests/translation.py:95]: But my base class is abstract?

Comment (benzkji) [reply modeltranslation/tests/translation.py:95]: Im my experience, it doesn't matter if the abstract or concrete model is registered. The problem is, that django thinks that the two fields of the child models belong to the same model, and then trigger the `TypeError / Cannot combine queries on two different base models` error, during the update query for the "models.SET_NULL" foreign keys. That's what I think, I might be wrong.

Comment (last-partizan) [reply modeltranslation/tests/translation.py:95]: Oh, i missed that part about abstract model. But anyway, you should not inherit options classes.

Comment (benzkji) [reply modeltranslation/tests/translation.py:95]: But that's what docs do/say? https://django-modeltranslation.readthedocs.io/en/latest/registration.html#translationoptions-fields-inheritance

Abstract base class: Inherit TranslationOptionsBaseClass. Concrete base class: Don't inherit TranslationOptions.

For me, it doesn't matter, and I will happily do what solves the issue (as you said, translate/register on the model where the field is defined), but for now there is no solution, or I did not proceed correctly. If you run tests, you'll see yourself.

Comment (benzkji) [reply modeltranslation/tests/translation.py:95]: And: Thank you for your help @last-partizan really appreciate!

File: modeltranslation/tests/translation.py (hunk 1 above)
Comment (benzkji) [reply modeltranslation/tests/translation.py]: None works.I've not found a solution, and provided a working (failing) test. So I guess it is a bug?

Comment (last-partizan) [reply modeltranslation/tests/translation.py]: Yeah, probably a bug.

**Comment by benzkji** [comment]
Any time/input, here? 

For me, it's a use case of modeltranslation - having an abstract base class with a foreign key, and having multiple classes that inherit from it, each using translated fields. As it's in the tests:
https://github.com/deschler/django-modeltranslation/pull/729/files#diff-a11e634034d805d4ce683d541f9b181dfc9e2d56beb6036134368586fba097d6R98

It's really deep in the Django core, I still think I was right here: "Django thinks that the two fields of the child models belong to the same model, and then trigger the TypeError / Cannot combine queries on two different base models error, during the update query for the "models.SET_NULL" foreign keys. That's what I think, I might be wrong."

So it's somewhere in Metaclasse, wherever hidden deep?

**Comment by last-partizan** [comment]
Hello!

Last time we stopped at the step where it fails just as it should fail, e.q. "issue is present and we have correct test-case for reproducing it". Right?

As for what to do next, I don't know.

Have you tried removing `TranslationOptions` here?

```py
class SecondSetNullForeignKeyModelTranslationOptions(SetNullForeignKeyBaseModelTranslationOptions, TranslationOptions):
    fields = [
        "translated22",
    ]
```

If removing it helps, great. If it doesn't help - try digging deeper, and ask questions. I'll try to share my knowledge, but this project is fairly complex and I just doing maintenance here, I'm not the original creator.
